pkg/controllers/uiplugin/controller.go

```diff
@@ -0,0 +1,278 @@
+
+	ctrl, err := ctrl.NewControllerManagedBy(mgr).
+		For(&uiv1alpha1.UIPlugin{}).
+		Owns(&appsv1.Deployment{}, generationChanged).
+		Owns(&v1.Service{}, generationChanged).
+		Owns(&v1.ServiceAccount{}, generationChanged).
+		Owns(&rbacv1.ClusterRole{}, generationChanged).
+		Owns(&rbacv1.ClusterRoleBinding{}, generationChanged).
+		Owns(&osv1alpha1.ConsolePlugin{}, generationChanged).
+		Build(rm)
+	if err != nil {
+		return err
+	}
+	rm.controller = ctrl
+
+	return nil
+}
+
+func getClusterVersion(k8client client.Reader) (*configv1.ClusterVersion, error) {
+	clusterVersion := &configv1.ClusterVersion{}
+	key := client.ObjectKey{Name: "version"}
+	if err := k8client.Get(context.TODO(), key, clusterVersion); err != nil {
+		return nil, err
+	}
+	return clusterVersion, nil
+}
+
+func (rm resourceManager) consolePluginCapabilityEnabled(ctx context.Context, name types.NamespacedName) bool {
+	current := &osv1alpha1.ConsolePlugin{}
+	err := rm.k8sClient.Get(ctx, name, current)
+
+	return err == nil || !metaerrors.IsNoMatchError(err)
+}
+
+func (rm resourceManager) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := rm.logger.WithValues("plugin", req.NamespacedName)
+
+	if !rm.consolePluginCapabilityEnabled(ctx, req.NamespacedName) {
+		logger.Info("Cluster console plugin not supported or not accesible. Skipping observability UI plugin reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	logger.Info("Reconciling observability UI plugin")
+
+	plugin, err := rm.getUIPlugin(ctx, req)
+	if err != nil {
+		// retry since some error has occured
+		return ctrl.Result{}, err
+	}
+
+	if plugin == nil {
+		// no such obs ui plugin, so stop here
+		return ctrl.Result{}, nil
+	}
+
+	if !plugin.ObjectMeta.DeletionTimestamp.IsZero() {
+		logger.V(6).Info("skipping reconcile since object is already schedule for deletion")
+		return ctrl.Result{}, nil
+	}
+
+	pluginInfo, err := PluginInfoBuilder(plugin, rm.pluginConf, rm.clusterVersion)
+
+	if err != nil {
+		logger.Error(err, "failed to reconcile plugin")
+		return ctrl.Result{}, err
+	}
+
+	reconcilers := pluginComponentReconcilers(plugin, *pluginInfo)
+	for _, reconciler := range reconcilers {
+		err := reconciler.Reconcile(ctx, rm.k8sClient, rm.scheme)
+		// handle creation / updation errors that can happen due to a stale cache by
+		// retrying after some time.
+		if apierrors.IsAlreadyExists(err) || apierrors.IsConflict(err) {
+			logger.V(8).Info("skipping reconcile error", "err", err)
+			return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+		}
+		if err != nil {
+			return rm.updateStatus(ctx, req, plugin, err), err
+		}
+	}
+
+	if err := rm.registerPluginWithConsole(ctx, pluginInfo); err != nil {
+		return rm.updateStatus(ctx, req, plugin, err), err
+	}
+
+	return rm.updateStatus(ctx, req, plugin, nil), nil
+}
+
+func (rm resourceManager) updateStatus(ctx context.Context, req ctrl.Request, pl *uiv1alpha1.UIPlugin, recError error) ctrl.Result {
```

Comment:
I see the status subresource being constantly updated (e.g. `lastTransitionTime` increases over time). It also triggers the reconciliation loop ever and ever.

Reply:
As discussed, this was caused by 2 instances trying to reconcile the same resources. I checked the kubebuilder validation, but wasn't able to find something that checks cross resources. It seems to run validation only on the current resource, so I'm inclined to write an admission webhook unless @jan--f knows a validation rule for this case.

Reply:
This is to validate that there is only a single UIplugin per cluster?

The trick for that is to add a validation to the yaml manifest (afaik this can't yet be added via a go comment, but we can use kustomize) that enforces a predefined name for the object and prevents any other object names.
See for example https://github.com/rhobs/observability-operator/compare/main...jan--f:observability-operator:config-crd#diff-9d14e1263c27af9a367892b64032baaa69bed686ae7e9c63cb5057fe6f3328eaR9-R26

Reply:
But as I said, I could see the being added in a later PR

Reply:
Multiple UIPugins can exists in a cluster, but only one per type (Dashboards, Logging, DistributedTracing, etc). I added a task: https://issues.redhat.com/browse/COO-122